# Model Algorithm 

#### Linear Regressor
#### K Neighbors Regressor
#### Decision Tree Regressor
#### Gradient boosting Regressor



In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
import warnings
warnings.filterwarnings('ignore')


In [2]:
df= pd.read_csv(r'C:\Users\user\Downloads\Scaled_Lower_Secondary (1).csv_')

In [3]:
df.head()

Year  Completion Rate for lower secondary Edu  Childhood Education GER  \
0  2012                                     99.6                47.379975   
1  2012                                     99.3                47.379975   
2  2013                                     99.6                47.379975   
3  2013                                     99.4                47.379975   
4  2014                                     99.6                36.650000   

   Gross enrolment ratio, early childhood educational development programmes  \
0                                           5.868056                           
1                                           5.868056                           
2                                           5.868056                           
3                                           5.868056                           
4                                           0.000000                           

   Gross intake ratio to the last grade of lower secondary general education,  \
0                                              96.64                            
1                                              97.66                            
2                                             103.13                            
3                                             102.95                            
4                                             105.74                            

   Literacy rate for 25-64 years old  \
0                             68.735   
1                             68.735   
2                             68.735   
3                             68.735   
4                             68.735   

   Expenditure on education as a percentage of total government expenditure (%)  \
0                                              19.25                              
1                                              19.25                              
2                                              17.96                              
3                                              17.96                              
4                                              17.32                              

   Government expenditure on education as a percentage of GDP (%)  \
0                                               3.90                
1                                               3.90                
2                                               3.44                
3                                               3.44                
4                                               3.42                

   Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
0           1.0                        0.0                             0.0   
1           1.0                        0.0                             0.0   
2           1.0                        0.0                             0.0   
3           1.0                        0.0                             0.0   
4           1.0                        0.0                             0.0   

   Europe and Northern America  Latin America and the Caribbean  \
0                          0.0                              0.0   
1                          0.0                              0.0   
2                          0.0                              0.0   
3                          0.0                              0.0   
4                          0.0                              0.0   

   Northern Africa and Western Asia  Oceania  Southern Asia  \
0                               0.0      0.0            0.0   
1                               0.0      0.0            0.0   
2                               0.0      0.0            0.0   
3                               0.0      0.0            0.0   
4                               0.0      0.0            0.0   

   Sub-Saharan Africa  Gender_numerical  
0                 0.0                 1  
1                 0.0                 0  
2                 0.0                 1 

## Data Transformation

#### 1. Data Rescaling

In [4]:
# Data Rescaling

data_col = df['Year']
df.set_index('Year',inplace =True)

df.head()


Completion Rate for lower secondary Edu  Childhood Education GER  \
Year                                                                     
2012                                     99.6                47.379975   
2012                                     99.3                47.379975   
2013                                     99.6                47.379975   
2013                                     99.4                47.379975   
2014                                     99.6                36.650000   

      Gross enrolment ratio, early childhood educational development programmes  \
Year                                                                              
2012                                           5.868056                           
2012                                           5.868056                           
2013                                           5.868056                           
2013                                           5.868056                           
2014                                           0.000000                           

      Gross intake ratio to the last grade of lower secondary general education,  \
Year                                                                               
2012                                              96.64                            
2012                                              97.66                            
2013                                             103.13                            
2013                                             102.95                            
2014                                             105.74                            

      Literacy rate for 25-64 years old  \
Year                                      
2012                             68.735   
2012                             68.735   
2013                             68.735   
2013                             68.735   
2014                             68.735   

      Expenditure on education as a percentage of total government expenditure (%)  \
Year                                                                                 
2012                                              19.25                              
2012                                              19.25                              
2013                                              17.96                              
2013                                              17.96                              
2014                                              17.32                              

      Government expenditure on education as a percentage of GDP (%)  \
Year                                                                   
2012                                               3.90                
2012                                               3.90                
2013                                               3.44                
2013                                               3.44                
2014                                               3.42                

      Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
Year                                                                            
2012           1.0                        0.0                             0.0   
2012           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2014           1.0                        0.0                             0.0   

      Europe and Northern America  Latin America and the Caribbean  \
Year                                                                 
2012                          0.0                              0.0   
2012                          0.0                              0.0   
2013                          0.0                         

In [5]:
scaler = MinMaxScaler()
normalized_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)
normalized_df['Year'] = data_col
normalized_df.set_index('Year', inplace =True)
normalized_df.head()

Completion Rate for lower secondary Edu  Childhood Education GER  \
Year                                                                     
2012                                 0.995830                 0.291370   
2012                                 0.992702                 0.291370   
2013                                 0.995830                 0.291370   
2013                                 0.993745                 0.291370   
2014                                 0.995830                 0.224366   

      Gross enrolment ratio, early childhood educational development programmes  \
Year                                                                              
2012                                           0.062733                           
2012                                           0.062733                           
2013                                           0.062733                           
2013                                           0.062733                           
2014                                           0.000000                           

      Gross intake ratio to the last grade of lower secondary general education,  \
Year                                                                               
2012                                           0.505572                            
2012                                           0.510908                            
2013                                           0.539524                            
2013                                           0.538582                            
2014                                           0.553178                            

      Literacy rate for 25-64 years old  \
Year                                      
2012                           0.657782   
2012                           0.657782   
2013                           0.657782   
2013                           0.657782   
2014                           0.657782   

      Expenditure on education as a percentage of total government expenditure (%)  \
Year                                                                                 
2012                                           0.482729                              
2012                                           0.482729                              
2013                                           0.446206                              
2013                                           0.446206                              
2014                                           0.428086                              

      Government expenditure on education as a percentage of GDP (%)  \
Year                                                                   
2012                                           0.286765                
2012                                           0.286765                
2013                                           0.252941                
2013                                           0.252941                
2014                                           0.251471                

      Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
Year                                                                            
2012           1.0                        0.0                             0.0   
2012           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2014           1.0                        0.0                             0.0   

      Europe and Northern America  Latin America and the Caribbean  \
Year                                                                 
2012                          0.0                              0.0   
2012                          0.0                              0.0   
2013                          0.0                         

In [6]:
normalized_df.corr()

Completion Rate for lower secondary Edu  \
Completion Rate for lower secondary Edu                                            1.000000   
Childhood Education GER                                                            0.343942   
Gross enrolment ratio, early childhood educatio...                                 0.328310   
Gross intake ratio to the last grade of lower s...                                 0.696288   
Literacy rate for 25-64 years old                                                  0.340289   
Expenditure on education as a percentage of tot...                                -0.045622   
Government expenditure on education as a percen...                                 0.186643   
Central Asia                                                                       0.201558   
Central and Southern Asia                                                          0.106305   
Eastern and South-Eastern Asia                                                     0.060688   
Europe and Northern America                                                        0.390367   
Latin America and the Caribbean                                                    0.157616   
Northern Africa and Western Asia                                                   0.081599   
Oceania                                                                            0.087659   
Southern Asia                                                                     -0.014274   
Sub-Saharan Africa                                                                -0.657521   
Gender_numerical                                                                   0.033615   

                                                    Childhood Education GER  \
Completion Rate for lower secondary Edu                            0.343942   
Childhood Education GER                                            1.000000   
Gross enrolment ratio, early childhood educatio...                 0.297850   
Gross intake ratio to the last grade of lower s...                 0.360958   
Literacy rate for 25-64 years old                                  0.118688   
Expenditure on education as a percentage of tot...                -0.179933   
Government expenditure on education as a percen...                -0.016032   
Central Asia                                                      -0.064723   
Central and Southern Asia                                          0.087554   
Eastern and South-Eastern Asia                                     0.157975   
Europe and Northern America                                        0.185838   
Latin America and the Caribbean                                    0.023716   
Northern Africa and Western Asia                                  -0.050977   
Oceania                                                            0.053288   
Southern Asia                                                      0.149977   
Sub-Saharan Africa                                                -0.346649   
Gender_numerical                                                   0.001953   

                                                    Gross enrolment ratio, early childhood educational development programmes  \
Completion Rate for lower secondary Edu                                                      0.328310                           
Childhood Education GER                                                                      0.297850                           
Gross enrolment ratio, early childhood educatio...                                           1.000000                           
Gross intake ratio to the last grade of lower s...                                           0.336662                           
Literacy rate for 25-64 years old                                                            0.122610                           
Expenditure on education as a percentage of tot...                                           0.003044                           
Government expe

In [8]:
# spliting the dataset into train set, test set, validation set.
datax = normalized_df.drop('Completion Rate for lower secondary Edu', axis = 1)
datay = normalized_df['Completion Rate for lower secondary Edu']

train_ratio = 0.75
val_ratio = 0.15
test_ratio = 0.10

# making the train set to be 75% of the entire dataset.
x_train, x_test, y_train, y_test = train_test_split(datax, datay, test_size = 1-train_ratio)

# test is now 10% of the entire dataset
# validation is now 15% of the entire dataset

x_val, x_test, y_val, y_test = train_test_split(x_test, y_test, test_size = test_ratio/(test_ratio + val_ratio))

print(x_train, x_val, x_test)

      Childhood Education GER  \
Year                            
2015                 0.170226   
2013                 0.291370   
2015                 0.291370   
2020                 0.692394   
2012                 0.353815   
...                       ...   
2020                 0.291370   
2012                 0.291370   
2018                 0.291370   
2014                 0.291370   
2016                 0.291370   

      Gross enrolment ratio, early childhood educational development programmes  \
Year                                                                              
2015                                           0.028330                           
2013                                           0.062733                           
2015                                           0.000000                           
2020                                           0.000000                           
2012                                           0.000000                 

## Model Training, Testing and evaluation

 
 Because the target variable been 'Completion rate for primary school Edu' is a continoues data and not categorical. therfore, the model algorithm to be use is the regression and not classifiers.

#### 1. Linear Regression


In [9]:

import time

# Record start time
start_time = time.time()

# Create a Linear Regression model
linear_model = LinearRegression()

# Fit the model to the training dataset
linear_model.fit(x_train, y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))





Training took 0.36 seconds.


In [10]:
# Obtain predictions
predicted_values = linear_model.predict(x_test)

#Measuring regression performance
MAE = mean_absolute_error(y_test, predicted_values)
MSE = mean_squared_error(y_test, predicted_values)
RMSE =np.sqrt(mean_squared_error(y_test, predicted_values))
RSS =np.sum(np.square(y_test, predicted_values))
r2=r2_score(y_test, predicted_values)




In [11]:
# Model evaluation

results = pd.DataFrame([['Linear regression', MAE, MSE, RMSE, RSS,r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results

Model       MAE       MSE      RMSE        RSS  r2_score
0  Linear regression  0.127095  0.025333  0.159163  99.987229  0.667829

In [12]:
lreg_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":predicted_values})
lreg_df


actual_values  predictions
Year                            
2016       0.800876     0.641402
2020       0.504796     0.254819
2020       0.919621     0.845702
2014       0.804003     0.646421
2016       0.968724     0.938426
...             ...          ...
2016       0.978107     0.956693
2016       0.834445     0.696299
2020       0.593411     0.352137
2017       0.979149     0.958733
2019       0.212886     0.045320

[188 rows x 2 columns]

#### Decition Tree Regressor

In [13]:
from sklearn.tree import DecisionTreeRegressor

import time


# Record start time
start_time = time.time()

DecisionTree_model = DecisionTreeRegressor()
DecisionTree_model.fit(x_train,y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))


Training took 0.13 seconds.


In [14]:
y_preds = DecisionTree_model.predict(x_test)

In [15]:
# Model Evaluation
mae= mean_absolute_error(y_test, y_preds)
mse= mean_squared_error(y_test, y_preds)
rmse=np.sqrt(mean_squared_error(y_test, y_preds))
rss=np.sum(np.square(y_test, y_preds))
r_squared=r2_score(y_test, y_preds)

In [16]:
model_results = pd.DataFrame([['Decision Tree regression', mae, mse, rmse, rss,r_squared]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [17]:
model_results

Model       MAE       MSE      RMSE        RSS  r2_score
0  Decision Tree regression  0.063537  0.013548  0.116396  99.987229  0.667829

In [18]:
Dreg_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":predicted_values})
Dreg_df


actual_values  predictions
Year                            
2016       0.800876     0.641402
2020       0.504796     0.254819
2020       0.919621     0.845702
2014       0.804003     0.646421
2016       0.968724     0.938426
...             ...          ...
2016       0.978107     0.956693
2016       0.834445     0.696299
2020       0.593411     0.352137
2017       0.979149     0.958733
2019       0.212886     0.045320

[188 rows x 2 columns]

#### K-Nearest Neighbors Regression (KNN Regression)

In [19]:

# Record start time
start_time = time.time()

knn = KNeighborsRegressor()

knn.fit(x_train, y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))




Training took 0.06 seconds.


In [20]:
knn_predictions  = knn.predict(x_test)

In [21]:
print(knn_predictions.shape, y_test.shape)

(188,) (188,)


In [22]:
# getting model scores
knn_r2 = r2_score(y_test, knn_predictions)
knn_mse = mean_squared_error( y_test,knn_predictions)
knn_rmse = np.sqrt(knn_mse)
knn_mae = mean_absolute_error(y_test,knn_predictions)
knn_rss = np.sum(np.square(y_test, knn_predictions))

In [23]:
model_results = pd.DataFrame([['KNN regression', knn_mae, knn_mse, knn_rmse, knn_rss,knn_r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [24]:
model_results

Model       MAE       MSE      RMSE        RSS  r2_score
0  KNN regression  0.077555  0.014458  0.120242  99.987229  0.821091

In [25]:
knn_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":knn_predictions})
knn_df.sort_index(ascending=True, inplace=True)
knn_df

actual_values  predictions
Year                            
2012       0.992702     0.985458
2012       0.836843     0.700307
2012       0.190992     0.036478
2012       0.320267     0.102571
2012       0.980192     0.960776
...             ...          ...
2020       0.482902     0.233195
2020       0.488115     0.238256
2020       0.309842     0.096002
2021       0.912010     0.831762
2021       0.992077     0.984216

[188 rows x 2 columns]

In [26]:
from sklearn.ensemble import GradientBoostingRegressor


# Record start time
start_time = time.time()

Gbr = GradientBoostingRegressor()
Gbr.fit(x_train,y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))



Training took 1.02 seconds.


In [27]:
Gbr_predictions  = Gbr.predict(x_test)

In [28]:
# getting model scores
Gbr_r2 = r2_score(y_test, Gbr_predictions)
Gbr_mse = mean_squared_error( y_test,Gbr_predictions)
Gbr_rmse = np.sqrt(Gbr_mse)
Gbr_mae = mean_absolute_error(y_test,Gbr_predictions)
Gbr_rss = np.sum(np.square(y_test, Gbr_predictions))

In [29]:
model_results = pd.DataFrame([['GBR regression', knn_mae, knn_mse, knn_rmse, knn_rss,knn_r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [30]:
model_results

Model       MAE       MSE      RMSE        RSS  r2_score
0  GBR regression  0.077555  0.014458  0.120242  99.987229  0.821091

In [31]:
Gbr_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":knn_predictions})
Gbr_df.sort_index(ascending=True, inplace=True)
Gbr_df

actual_values  predictions
Year                            
2012       0.992702     0.985458
2012       0.836843     0.700307
2012       0.190992     0.036478
2012       0.320267     0.102571
2012       0.980192     0.960776
...             ...          ...
2020       0.482902     0.233195
2020       0.488115     0.238256
2020       0.309842     0.096002
2021       0.912010     0.831762
2021       0.992077     0.984216

[188 rows x 2 columns]

In [32]:
# Printing the stored results of all the three models
results.drop_duplicates()


Model       MAE       MSE      RMSE        RSS  r2_score
0         Linear regression  0.127095  0.025333  0.159163  99.987229  0.667829
1  Decision Tree regression  0.063537  0.013548  0.116396  99.987229  0.667829
2            KNN regression  0.077555  0.014458  0.120242  99.987229  0.821091
3            GBR regression  0.077555  0.014458  0.120242  99.987229  0.821091

### conclusion

1.Linear Regression time for the model to train is 0.36 seconds

1.Decision Tree Regression seems to perform well in terms of minimizing errors and time for the model to tarin is 0.13 seconds.

2.KNN Regression and GBR Regression have higher R-squared values, suggesting better explanatory power.
also, the time for the two models to train are 0.06 and 1.02 seconds respectively.

3.Both KNN and GBR have the same MAE, MSE, and RMSE, indicating similar performance on average.